external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(not(feature = "gas-profiler"))]
+pub(crate) use vm_tracer_noop::*;
+mod vm_tracer_noop {
+    use move_trace_format::format::MoveTraceBuilder;
+
+    #[allow(dead_code)]
+    pub(crate) struct VMTracer<'a> {
+        _phantom: std::marker::PhantomData<&'a ()>,
+    }
+
+    impl<'a> VMTracer<'a> {
+        #[allow(dead_code)]
+        pub(crate) fn new(_trace: &'a mut MoveTraceBuilder) -> Self {
+            Self {
+                _phantom: std::marker::PhantomData,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "gas-profiler")]
+pub(crate) use vm_tracer_impl::*;
+#[cfg(feature = "gas-profiler")]
+mod vm_tracer_impl {
+    use crate::{
+        interpreter::{Frame, Interpreter},
+        loader::{Function, Loader},
+    };
+    use move_binary_format::{
+        errors::PartialVMError,
+        file_format::{ConstantPoolIndex, SignatureIndex},
+    };
+    use move_core_types::{
+        account_address::AccountAddress,
+        annotated_value::{MoveTypeLayout, MoveValue},
+        language_storage::TypeTag,
+    };
+    use move_trace_format::format::{
+        DataLoad, Effect as EF, Location, MoveTraceBuilder, Read, RefType, TraceIndex, TraceValue,
+        TypeTagWithRefs, Write,
+    };
+    use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+    use std::collections::BTreeMap;
+
+    /// Internal state for the tracer. This is where the actual tracing logic is implemented.
+    pub(crate) struct VMTracer<'a> {
+        trace: &'a mut MoveTraceBuilder,
+        link_context: Option<AccountAddress>,
```

Comment:
We need this in the tracer so that can get the correct module to e.g., resolve type layouts.

In the future, this can be replaced by the mapping that we get from the link context and/or the resolver over the package vtables. But at least for now/in the current incarnation of the VM here we need this. (note that this isn't exposed in the trace or anything though!)